File: dtrules/src/main/java/com/dtrules/decisiontables/CNode.java
/** 
 * Copyright 2004-2011 DTRules.com, Inc.
 * 
 * See http://DTRules.com for updates and documentation for the DTRules Rules Engine  
 *   
 * Licensed under the Apache License, Version 2.0 (the "License");  
 * you may not use this file except in compliance with the License.  
 * You may obtain a copy of the License at  
 *   
 *      http://www.apache.org/licenses/LICENSE-2.0  
 *   
 * Unless required by applicable law or agreed to in writing, software  
 * distributed under the License is distributed on an "AS IS" BASIS,  
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
 * See the License for the specific language governing permissions and  
 * limitations under the License.  
 **/

  
package com.dtrules.decisiontables;

import com.dtrules.infrastructure.RulesException;
import com.dtrules.interpreter.IRObject;
import com.dtrules.session.DTState;
/**
 * The Condition Node evaluates conditions within a Decision
 * table.
 * @author Paul Snow
 *
 */
public class CNode implements DTNode {

    final int            column;                  //Column that created this node
    final int            conditionNumber;         //NOPMD 
    final IRObject       condition;
    final RDecisionTable dtable;                  // Pointer back to the Decision Table for debug purposes
    boolean              star;
    
    DTNode iftrue  = null;
    DTNode iffalse = null;

    
    
    /**
     * Clone this CNode and all the CNodes referenced by this CNode
     */
    public CNode cloneDTNode (){
        CNode newCNode = new CNode(dtable,column,conditionNumber,condition);
        newCNode.iftrue  = iftrue  == null ? null : iftrue.cloneDTNode(); 
        newCNode.iffalse = iffalse == null ? null : iffalse.cloneDTNode();
        return newCNode;
    }
    
    public int getRow(){ return conditionNumber; }
    /**
     * Two CNodes are equal if their true paths and their
     * false paths are the same.
     * And those CommonANodes have to be equal.
     */
	public boolean equalsNode(DTState state, DTNode node) {
        if(node instanceof CNode){
            CNode cnode = (CNode)node;
            if(cnode.conditionNumber != conditionNumber){
                return false;
            }
            return(cnode.iffalse.equalsNode(state, iffalse) && 
               cnode.iftrue.equalsNode(state, iftrue));
        }else{
            ANode me  = getCommonANode(state);  // Get this CNode's commonANode.
            if(me==null)return false;           // If none exists, it can't be equal!
            ANode him = getCommonANode(state);  // Get the other DTNode's commonANode
            if(him==null)return false;          // If none exists, it can't be equal!
            return me.equalsNode(state, him);   // Return true if this node matches that node!
        }    
    }
	
    /**
     * To have a CommonANode, every path through the CNode (i.e. both
     * the true path and the false path) has to have the same commonANode.
     * So both iftrue and iffalse have to have a commonANode, and those have
     * to match.
     */
    public ANode getCommonANode(DTState state) {
        if(state.testState(DTState.TRACE) &&                // If we are tracing, don't do this.
                iftrue.equalsNode(state, iffalse)){         // If both true/false paths are the same, 
            ANode trueSide  = iftrue.getCommonANode(state); //   then we will merge them.  Get the 
            ANode falseSide = iffalse.getCommonANode(state);
            trueSide.addNode(falseSide);                    
            if(trueSide instanceof ANode){                  // We add them together so we maintain the tracking
                ((ANode) trueSide).addNode((ANode)falseSide);   //   of columns tested.
            }
            return trueSide;  
        }
        return null;                                        // If they don't match, I have to return false!
    }

  
	
	CNode(RDecisionTable dt, int column, int conditionNumber, IRObject condition){
        this.column          = column;
        this.conditionNumber = conditionNumber; 
        this.condition       = condition;
        this.dtable          = dt;
    }
	
	public int countColumns(){
	    int t = iftrue.countColumns();
	    int f = iffalse.countColumns();
	    return t+f;
	}
	    
	public void execute(DTState state) throws RulesException{
        boolean result;
        try {
            if(state.testState(DTState.TRACE)){
            	if(state.testState(DTState.VERBOSE)){
	            	state.traceTagBegin("Condition", "n",(conditionNumber+1)+"");
	                    state.traceInfo("Formal", dtable.getConditions()[conditionNumber]);
	                    state.traceInfo("Postfix", dtable.getConditionsPostfix()[conditionNumber]);
	                    state.traceTagBegin("execute");
	                    result = state.evaluateCondition(condition);
	                    state.traceTagEnd();
	                    state.traceInfo("result", "v",(result?"Y":"N"),null);
	                state.traceTagEnd();
            	}else{
            		 result = state.evaluateCondition(condition);
            		 state.traceInfo("Condition", "n",(conditionNumber+1)+"","v",(result?"Y":"N"),null);
            	}
            }else{
                result = state.evaluateCondition(condition);
            }    		
    		
        } catch (RulesException e) {
            e.isFirstAction();  // Avoids a bogus "first action"
            if(state.testState(DTState.TRACE)){
                state.traceTagEnd();
                state.traceInfo("result", "v", "ERROR",null);
                state.traceTagEnd();
            }
            e.setSection("Condition",conditionNumber+1);
            e.setFormal(dtable.getConditions()[conditionNumber]);
            throw e;
        } catch (Exception e){
            RulesException re = new RulesException(e.getClass().getName(), e.getStackTrace()[0].getClassName(), e.getMessage());
            re.isFirstAction();  // Avoids a bogus "first action"
            if(state.testState(DTState.TRACE)){
                state.traceTagEnd();
                state.traceInfo("result", "v", "ERROR",null);
                state.traceTagEnd();
            }
            re.setSection("Condition",conditionNumber+1);
            re.setFormal(dtable.getConditions()[conditionNumber]);
            throw re;
        }

        if(result){
            iftrue.execute(state);
        }else{
            iffalse.execute(state);
        }

	}

	public Coordinate validate() {
		if(iftrue  == null || iffalse == null){
           return new Coordinate(conditionNumber,column); 
        }   
	    Coordinate result = iffalse.validate();
        if(result!=null){
            return result;
        }
        
        return iftrue.validate();
	}

	public String toString(){
		return "Condition Number "+(conditionNumber+1);
	}
	
	public void addNode(DTNode _node) {
	    if(!(_node instanceof CNode)){
	        throw new RuntimeException("Shouldn't ever attempt to combine nodes of different types");
	    }
	    CNode node = (CNode) _node;
	    iftrue.addNode(node.iftrue);
	    iffalse.addNode(node.iffalse);
	}

    public boolean getStar() {
        return star;
    }

    public void setStar(boolean star) {
        this.star = star;
    }
	
}

File: pkg/dtrules/decisiontable/cnode.go
// Copyright 2004-2011 DTRules.com, Inc.
// Copyright 2024 Paul Snow
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package decisiontable

import (
	"fmt"

	"github.com/DTRules/DTRules/pkg/dtrules"
)

// CNode is a Condition Node that evaluates conditions within a Decision Table.
// It forms the internal nodes of the decision tree.
type CNode struct {
	column          int                    // Column that created this node
	conditionNumber int                    // Row number (0-based) of the condition
	condition       dtrules.Object         // The compiled condition code (Go interpreter)
	bytecode        *dtrules.BytecodeChunk // Compiled bytecode (for ASM execution)
	decisionTable   *RDecisionTable        // Pointer back to the Decision Table
	star            bool                   // Whether this column has a star

	IfTrue  DTNode // Branch taken if condition is true
	IfFalse DTNode // Branch taken if condition is false
}

// SetBytecode sets the bytecode for this condition.
func (c *CNode) SetBytecode(bc *dtrules.BytecodeChunk) {
	c.bytecode = bc
}

// GetBytecode returns the bytecode for this condition.
func (c *CNode) GetBytecode() *dtrules.BytecodeChunk {
	return c.bytecode
}

// NewCNode creates a new condition node
func NewCNode(dt *RDecisionTable, column, conditionNumber int, condition dtrules.Object) *CNode {
	return &CNode{
		column:          column,
		conditionNumber: conditionNumber,
		condition:       condition,
		decisionTable:   dt,
	}
}

// GetRow returns the condition row number (0-based)
func (c *CNode) GetRow() int {
	return c.conditionNumber
}

// CloneDTNode creates a deep copy of this CNode and all its sub-nodes
func (c *CNode) CloneDTNode() DTNode {
	newNode := &CNode{
		column:          c.column,
		conditionNumber: c.conditionNumber,
		condition:       c.condition,
		decisionTable:   c.decisionTable,
		star:            c.star,
	}
	if c.IfTrue != nil {
		newNode.IfTrue = c.IfTrue.CloneDTNode()
	}
	if c.IfFalse != nil {
		newNode.IfFalse = c.IfFalse.CloneDTNode()
	}
	return newNode
}

// CountColumns returns the total number of columns (paths) through this node
func (c *CNode) CountColumns() int {
	t := 0
	f := 0
	if c.IfTrue != nil {
		t = c.IfTrue.CountColumns()
	}
	if c.IfFalse != nil {
		f = c.IfFalse.CountColumns()
	}
	return t + f
}

// Execute evaluates the condition and follows the appropriate branch
func (c *CNode) Execute(state dtrules.State) error {
	// Evaluate the condition
	result, err := state.EvaluateCondition(c.condition)
	if err != nil {
		return fmt.Errorf("condition %d in table %s: %w",
			c.conditionNumber+1, c.decisionTable.GetName(), err)
	}

	// Trace: record which condition ran and how it evaluated, so a loaded
	// trace can annotate the grid with actual results. Guarded: the Sprintf
	// arguments were ~8% of a table execution's CPU with tracing off (#1025).
	if state.Tracing() {
		state.TraceInfo("condition", "",
			"n", fmt.Sprintf("%d", c.conditionNumber+1),
			"result", fmt.Sprintf("%t", result))
	}

	// Follow the appropriate branch
	if result {
		if c.IfTrue != nil {
			return c.IfTrue.Execute(state)
		}
	} else {
		if c.IfFalse != nil {
			return c.IfFalse.Execute(state)
		}
	}
	return nil
}

// Validate checks that both branches are properly connected
func (c *CNode) Validate() *Coordinate {
	if c.IfTrue == nil || c.IfFalse == nil {
		return &Coordinate{Row: c.conditionNumber, Col: c.column}
	}

	result := c.IfFalse.Validate()
	if result != nil {
		return result
	}

	return c.IfTrue.Validate()
}

// EqualsNode returns true if both nodes execute the same actions for all paths
func (c *CNode) EqualsNode(state dtrules.State, node DTNode) bool {
	if other, ok := node.(*CNode); ok {
		if other.conditionNumber != c.conditionNumber {
			return false
		}
		return other.IfFalse.EqualsNode(state, c.IfFalse) &&
			other.IfTrue.EqualsNode(state, c.IfTrue)
	}

	// Compare with an ANode by getting common paths
	me := c.GetCommonANode(state)
	if me == nil {
		return false
	}
	him := node.GetCommonANode(state)
	if him == nil {
		return false
	}
	return me.EqualsNode(state, him)
}

// GetCommonANode returns an ANode if both branches lead to the same actions
func (c *CNode) GetCommonANode(state dtrules.State) *ANode {
	// Both paths must lead to the same result
	if !c.IfTrue.EqualsNode(state, c.IfFalse) {
		return nil
	}

	trueSide := c.IfTrue.GetCommonANode(state)
	falseSide := c.IfFalse.GetCommonANode(state)
	if trueSide == nil || falseSide == nil {
		return nil
	}

	// Merge the two sides (for column tracking)
	// Ignore error here as we've already verified they're compatible
	_ = trueSide.AddNode(falseSide)
	return trueSide
}

// AddNode combines another node into this node
// Returns error if the nodes are incompatible types
func (c *CNode) AddNode(node DTNode) error {
	other, ok := node.(*CNode)
	if !ok {
		return dtrules.TypeCheckError("CNode.AddNode", "cannot combine CNode with different node type")
	}
	if err := c.IfTrue.AddNode(other.IfTrue); err != nil {
		return err
	}
	return c.IfFalse.AddNode(other.IfFalse)
}

// GetStar returns whether this column has a star
func (c *CNode) GetStar() bool {
	return c.star
}

// SetStar sets the star flag
func (c *CNode) SetStar(star bool) {
	c.star = star
}

// GetConditionNumber returns the condition row number (0-based)
func (c *CNode) GetConditionNumber() int {
	return c.conditionNumber
}

// GetColumn returns the column that created this node
func (c *CNode) GetColumn() int {
	return c.column
}

// String returns a string representation of this CNode
func (c *CNode) String() string {
	return fmt.Sprintf("Condition Number %d", c.conditionNumber+1)
}
